json/DataTypes/Decoding.roc

```diff
@@ -0,0 +1,167 @@
+module [string, number, bool, list, field, map, map2, map3, andThen, or, JsonDecoder, DecodingErrors]
+import JsonData exposing [Json]
+
+DecodingErrors : [
+    FieldNotFound Str,
+    ExpectedJsonObject Str,
+    ExpectedJsonArray Str,
+    WrongJsonType Str,
+    KeyNotFound,
+    DecodingFailed
+]
+    
+JsonDecoder t : Json -> Result t DecodingErrors
```

Comment:
I'm not sure if you want to get into it in this chapter, but an even more idiomatic way to do this would be to make `JsonDecoder` be an [opaque type](https://www.roc-lang.org/tutorial#opaque-types) which wraps a function instead of a type alias for a function.

It wouldn't be much of a structural change to the code - you'd use `JsonDecoder t :=` instead of `JsonDecoder t :` here, and then things like `string = \@JsonDecoder json ->` in a bunch of places, and then expose a function to actually run the decoder like:

```roc
decode : JsonDecoder t, Json -> Result t Decoding Error
decode = \@JsonDecoder fn, json -> fn json
```

...which other modules would call instead of passing `json` directly to the decoder, because it would be a wrapper around a function instead of a function.

This is the more idiomatic way because it means things like:
- Once it's an opaque type, you can customize what Abilities the decoder has, and how they work (`Eq`, `Hash`, etc.)
- You can change its internal structure without the possibility of breaking callers, because they can't access its internal structure directly anymore.

That said, I think it's also fine to consider that too much scope for a chapter that's mainly about JSON decoding. You could consider mentioning it at the end - like "to make this really idiomatic, we'd make `JsonDecoder` into an opaque type. You can read more about opaque types in Chapter ____.

Reply:
I like the sound of making the decoder an opaque type, and I will probably do this for my benefit and understanding. 

As for including it in the chapter, I am unsure if it'll be too much to go into it or not. I think it'll become a lot clearer when I go to write the content properly.